### PR TITLE
Az167 add pr to get fsm + more cleanups.

### PR DESCRIPTION
**Pull request by jonmeredith**
Refactors the get FSM 
- Replaces r and timeout arguments with options instead.
- Adds pr option to supply a minimum number of primary nodes in the preflist (but does not guarantee responses from them).
- Adds basic_quorum option to control whether basic_quorum is enabled.
- Adds notfound_ok option to count notfound responses towards R.
- Adds details option to get additional information back from the FSM.
- Changed behavior when R is not met to return an {error, {r_val_unsatisfied, R, Got}} message rather than not_found.

<pre>
I'm about to add some features to the get and put FSMs needed by Trifork / Netic for that address the availability requirements for their presciption tracking application.

The big things are
1) Add PR parameter to get FSM to ensure PR vnode owners are involved (was RP during
   earlier discussion but I think PR reads better and matches W/DW model)
2) Add PW parameter to put FSM to ensure PW vnode owners are involved (again, was WP before)
3) Control counting not_found vnode responses towards R.
4) Return R value not satisfied as {error, {r_value_unsatisfied, R, P}} rather than notfound and pass through PBC/HTTP interfaces.

Rather than adding new arguments as we currently do and now we have the quorum information in the bucket props I think we should change the interface to get / put.

The current interfaces of
  riak_kv_get_fsm:start_link(ReqId,Bucket,Key,R,Timeout,From).
  riak_kv_put_fsm:start_link(ReqId,RObj,W,DW,Timeout,From, Options)

Would become
  riak_kv_get_fsm:start_link(From, Bucket, Key, GetOptions) -> {ok, Pid}
  riak_kv_get_fsm:start_link(From, Object, PutOptions) -> {ok, Pid}

  From = {raw, ReqId, Pid}.   %% Reuse the format used by riak_core_vnode.

  GetOptions = [{r, R},
                {pr, PR},
                {notfound_ok, bool()}, % default false, flip in future release
                {timeout, Timeout},
                {details, Details}].

  PutOptions = [{w, DW},
                {dw, DW},
                {pw, PW},
                {timeout, Timeout},
                {details, Details},
                {returnbody, bool()}].

  Details = false | true | [timing, bestnode, primaries, 
                            params, vnodes]

If the details option is set true we'll return an extra element to the tuple with
a proplist giving all details about the request, individual details can be requested
by grouping and if set false we just return normally.

If timing set, add entries for internal timing
  {duration, msecs()}

If bestnode set, return the best node for resubmitting future requests to (initially will be first responder)
  {bestnode, node()}

If primaries set, return the count of primary nodes *successfully* participating in the request.
  {primaries, integer()}

If params set, add entries for FSM parameters (after decoding symbolic names)
  n, r, pr, w, dw, pw

If vnodes set, add entries
  {node_oks, Count}, {node_errors, [Reason]}

The response from get would become {ok, riak_object(), info()} or {error, Reason, info()} - e.g. {error, notfound, info()}
Similarly for put we will return {ok, info()} or {ok, riak_object, info()} or {error, Reason, info()}.

The new interface makes r / w / dw etc optional - so we set as default the way riak_client:put does currently.
</pre>

### CODE REVIEW

**Comment by justinsheehy** [comment]
I went through all of these commits.

Not only does the code appear to work as it should, it's also clear and easy to read.

+1

File: src/riak_client.erl [hunk 1]
```diff
@@ -220,7 +220,24 @@ mapred_dynamic_inputs_stream(FSMPid, InputDef, Timeout) ->
 %% @doc Fetch the object at Bucket/Key.  Return a value as soon as the default
 %%      R-value for the nodes have responded with a value or error.
 %% @equiv get(Bucket, Key, R, default_timeout())
-get(Bucket, Key) -> get(Bucket, Key, default, ?DEFAULT_TIMEOUT).
+get(Bucket, Key) -> 
+    get(Bucket, Key, []).
+
+%% @spec get(riak_object:bucket(), riak_object:key(), options()) ->
```
Comment (argv0) [inline src/riak_client.erl:226]: maybe some edoc with the valid options here so people don't have to go digging into the FSM?

**Comment by slfritchie** [comment]
Jon, could you take a peek into this error?

# 

## FAIL: test_javascript_bucket_map_reduce (riak.tests.test_all.RiakPbcTransportTestCase)

Traceback (most recent call last):
  File "/Users/fritchie/b/src/riak-python-client/riak/tests/test_all.py", line 283, in test_javascript_bucket_map_reduce
    self.assertEqual(result, [9])
AssertionError: [7] != [9]

---

Ran 45 tests in 3.615s

FAILED (failures=1)

... triggered by basho_expect's python tests. A shortcut to run them would
be something like:

sh -c "cd /path/to/src/riak-python-client && env SKIP_SEARCH=1 RIAK_TEST_HOST=10.1.1.74 RIAK_TEST_HTTP_PORT=8098 RIAK_TEST_PB_PORT=8087 ./setup.py test"

**Comment by slfritchie** [comment]
Oops, I forgot to mention that, at Andrew's suggestion, some Protocol
Buffers errors need a workaround by "git rebase master" on top of your
az167-add-pr-to-get-fsm branch.

**Comment by krestenkrab** [comment]
So, how do I get the semantics I requested?  is that default, or do I have to use notfound_ok or basic_quorum options?  Argv0 is right, some basic doc would be good.

**Comment by jonmeredith** [comment]
Agreed on the docs.

We'll go through and enhance the docs when we add the last few options planned and do final cleanups. I've got a few more details I'd like to add and I'd like to make it possible to configure the defaults per-bucket.

For now you can get your desired semantics by using options of [{pr, 2}, {r, 2}, {notfound_ok, true}, {basic_quorum, false}]. 
That will require 2 primaries to participate in the preference list, 2 successful responses, counting notfound as successful and disable the basic quorum logic.

The defaults currently behave like the 0.14.1 release -- we'll discuss how to set them before or next release to minimize surprise to upgraders.

**Comment by slfritchie** [comment]
Jon, after a "git rebase master" on top of your branch in order to get all
of the basho_expect tests to stop complaining, I see the following Dialyzer
warnings which I believe are new (after Andy's big cleanup a few days ago).
Would you mind fixing them up?

```
riak_kv_get_fsm.erl:145: Matching of pattern <'timeout', StateData = {'state', {'raw', ReqId, _Pid}, Options, N, _, _, _, _, PL2, _, _, _, _, _, _, _, _, _, _, _, _, BucketProps, _, _}> tagged with a record name violates the declared type of #state{from::'undefined' | {integer(),pid()},options::[{'r',pos_integer()} | {'timeout','infinity' | pos_integer()}],n::'undefined' | pos_integer(),r::'undefined' | pos_integer(),fail_threshold::'undefined' | pos_integer(),allowmult::'false' | 'true' | 'undefined',notfound_ok::'false' | 'true' | 'undefined',preflist2::'undefined' | [{{non_neg_integer(),atom()},'fallback' | 'primary'}],req_id::'undefined' | pos_integer(),starttime::'undefined' | pos_integer(),replied_r::[any()],replied_notfound::[any()],replied_fail::[any()],final_obj::'tombstone' | 'undefined' | {'error','notfound'} | {'ok',riak_object:riak_object()},timeout::'infinity' | 'undefined' | pos_integer(),tref::'undefined' | reference(),bkey::'undefined' | {binary(),binary()},startnow::'undefined' | {pos_integer(),pos_integer(),pos_integer()},get_usecs::'undefined' | non_neg_integer()}
riak_kv_get_fsm.erl:152: Function will never be called
riak_kv_get_fsm.erl:365: Function get_option/3 will never be called
riak_kv_get_fsm.erl:373: Matching of pattern <Reply, StateData = {'state', {'raw', ReqId, Pid}, Options, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _}> tagged with a record name violates the declared type of #state{from::'undefined' | {integer(),pid()},options::[{'r',pos_integer()} | {'timeout','infinity' | pos_integer()}],n::'undefined' | pos_integer(),r::'undefined' | pos_integer(),fail_threshold::'undefined' | pos_integer(),allowmult::'false' | 'true' | 'undefined',notfound_ok::'false' | 'true' | 'undefined',preflist2::'undefined' | [{{non_neg_integer(),atom()},'fallback' | 'primary'}],req_id::'undefined' | pos_integer(),starttime::'undefined' | pos_integer(),replied_r::[any()],replied_notfound::[any()],replied_fail::[any()],final_obj::'tombstone' | 'undefined' | {'error','notfound'} | {'ok',riak_object:riak_object()},timeout::'infinity' | 'undefined' | pos_integer(),tref::'undefined' | reference(),bkey::'undefined' | {binary(),binary()},startnow::'undefined' | {pos_integer(),pos_integer(),pos_integer()},get_usecs::'undefined' | non_neg_integer()}
riak_kv_get_fsm.erl:400: The variable X can never match since previous clauses completely covered the type 'tombstone' | {'error','notfound'} | {'ok',riak_object:riak_object()}
riak_kv_get_fsm.erl:434: Function client_info/3 will never be called
riak_kv_get_fsm.erl:447: Function will never be called
```

**Comment by slfritchie** [comment]
Please ignore my earlier comments about basho_expect, python & PHP
client errors, they were results of my bad patch build & testing environment,
sorry!

I've a comment for types added/edited before your branch started:
- should fail_threshold's type be non_neg_integer()?
- the 3 elements of startnow should be non_neg_integer().

Zap a couple of stray blank lines, one in execute() and one in
waiting_vnode_r() (both in riak_kv_get_fsm.erl)

If you're confident that the sleep in get_fsm_qc.erl isn't necessary,
then please remove:

```
    % Give read repairs and deletes a chance to go through
    %timer:sleep(5),
```

The Dialyzer comments from earlier today ... and then it'll all look even better.
The improvements in readability of both the get FSM and its QuickCheck test
is very nice.

**Comment by krestenkrab** [comment]
For the changes needed to support the rest API, [Pull 39](https://github.com/basho/riak_kv/pull/39/files#diff-2) is a starting point; that code at least understands the new error values. But it will also need some code to be able to pass in the new parameters. 

Have you considered making some of the parameters configurable as bucket properties?  N is, anyway...

**Comment by jonmeredith** [comment]
Hi Kresten, yes - I'll make them bucket properties too - I just wanted the main code available for review and so the work on adding to the HTTP/PBC interfaces can proceed.
We've reviewed pull 39 as the basis for the HTTP changes - many thanks for submitting it.

**Comment by slfritchie** [comment]
Looks good to merge, thanks!

**Comment by slfritchie** [comment]
Looks good to merge, thanks!

**Comment by slfritchie** [comment]
Looks good to merge, thanks!

**Comment by krestenkrab** [comment]
Hooray!

**Pull request merged**